resource/move_state.go

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package resource
+
+import (
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-framework/internal/privatestate"
+	"github.com/hashicorp/terraform-plugin-framework/tfsdk"
+	"github.com/hashicorp/terraform-plugin-go/tfprotov6"
+)
+
+// MoveStateRequest represents a request for the provider to move a source
+// resource state into the target resource state with any necessary data
+// transformation logic. An instance of this request struct is supplied as an
+// argument to each [StateMover.StateMover].
+type MoveStateRequest struct {
+	// SourcePrivate is the source resource private state data. If this source
+	// data is important for the target resource, the implementation should set
+	// the data via [MoveStateResponse.TargetPrivate] as it is intentionally not
+	// copied automatically.
+	SourcePrivate *privatestate.ProviderData
+
+	// SourceProviderAddress is the address of the provider for the source
+	// resource type. It is the full address in HOSTNAME/NAMESPACE/TYPE format.
+	// For example, registry.terraform.io/hashicorp/random.
+	//
+	// Implementations should consider using this value to determine whether the
+	// request should be handled by this particular implementation. It is
+	// recommended to ignore the hostname unless necessary for disambiguation.
+	SourceProviderAddress string
+
+	// SourceRawState is the raw state of the source resource. This data is
+	// always available, regardless whether the [StateMover.SourceSchema] field
+	// was set. If SourceSchema is present, the [SourceState] field will be
+	// populated and it is recommended to use that field instead.
+	//
+	// If this request matches the intended implementation, the implementation
+	// logic must set [MoveStateResponse.State] as it is intentionally not
+	// copied automatically.
+	//
+	// This is advanced functionality for providers wanting to skip the full
+	// redeclaration of source schemas and instead use lower level handlers to
+	// transform data. A typical implementation for working with this data will
+	// call the Unmarshal() method.
+	//
+	// TODO: Create framework defined type that is not protocol specific.
+	// Reference: https://github.com/hashicorp/terraform-plugin-framework/issues/340
+	SourceRawState *tfprotov6.RawState
```

Comment:
I mentioned this in another comment, but this issue is closed now. Do we need to re-open it or remove this comment?